### PR TITLE
Coroutine support!

### PR DESCRIPTION
**Pull request by tgoodlet**
This adds initial support for coroutine based apps.

I've dropped callback support for now (debatable whether we should keep it) and this is demonstrated in a couple apps, namely the `CDR` app and a simple bridger in the test suite.

Looking for harsh criticism of both the design and API ;)

Thanks guys!

### CODE REVIEW

File: switchio/apps/measure/cdr.py [hunk 1]
```diff
@@ -154,60 +157,77 @@ def on_create(self, sess):
...
-            calleetimes['create'] if callee else None,
-            calleetimes['answer'] if callee else None,
-            calleetimes['hangup'] if callee else None,
-            pool.count_failed(),
-            call.vars['session_count'],
-            call.vars['erlangs'],
-        ))
-        if rollover:
-            self.log.debug('wrote data to disk')
+        events_sub = list(self.on_create.switchio_events_sub)
```
Comment (tgoodlet) [inline switchio/apps/measure/cdr.py]: Yes I realize this code is actually longer - so this is an example of where callbacks might be better suited anyway unless you guys can think of a better way to write this.

**Comment by tgoodlet** [comment]
Tests are currently failing due to some apps still using callbacks and a failed speed test likely because `uvloop` isn't required by default. Maybe it should be?

File: switchio/apps/measure/cdr.py (hunk 1 above)
Comment (vodik) [reply switchio/apps/measure/cdr.py]: I don't prevent you from maintaining the previous API? Why not support both?

And maybe you can look at supporting `async for` for python 3.6 code:

```python
async for event in await sessioon:
    # ...
```

Comment (tgoodlet) [reply switchio/apps/measure/cdr.py]: @vodik ok cool. I love the idea of `async for` that's super cool.

File: switchio/loops/proactor.py [hunk 2]
```diff
@@ -110,6 +117,84 @@ def start(self):
...
+                cid = model.cid if model else self.get_id(e, 'default')
+                self.log.debug("consumers id is '{}'".format(cid))
+
+                if model:
+                    # signal any awaiting futures
+                    fut = model._futures.get(evname, None)
+                    if fut:
+                        fut.set_result(e)
+                        # seriously guys, this is literally so stupid
+                        await asyncio.sleep(0)  # resume waiting coroutines...
```
Comment (tgoodlet) [inline switchio/loops/proactor.py]: @vodik you ever look at the [stdlib's impl](https://github.com/python/cpython/blob/master/Lib/asyncio/tasks.py#L476) for this?
Seems to me there should be some explicit way to resume all coroutines waiting on a task/future.

Do you guys have a more proper way in `aiosip`?

File: switchio/apps/measure/cdr.py (hunk 1 above)
Comment (tgoodlet) [reply switchio/apps/measure/cdr.py]: @vodik yeah only thing I'm thinking right now is that adding `async for` support pushes out `pypy` from the mix. You mind making an issue explaining the API you'd like to see?

File: switchio/loops/proactor.py (hunk 2 above)
Comment (tgoodlet) [reply switchio/loops/proactor.py]: Heh so looking into this apparently it's a common quiff with `asyncio`.
This [SO answer](https://stackoverflow.com/questions/29868372/python-asyncio-run-event-loop-once/29868627#29868627) claims the correct solution is to stop and start the loop...

**Comment by tgoodlet** [comment]
@vodik @moises-silva there that keeps in py2 support for the first release and adds back the callback API.

I think that should be it unless you guys have further suggestions.

File: switchio/loops/proactor.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+@asyncio.coroutine
+def just_yield():
+    """A "just yield" coroutine which triggers an interation of the event loop.
+
+    If you think this is a nightmare to understand have you asked yourself how
+    this will ever work once these legacy types of generator "coroutines" are
+    removed from the language?
+    """
+    yield
```
Comment (vodik) [inline switchio/loops/proactor.py]: The approach I took in some test code that needed to do this:

```python
release = functools.partial(asyncio.sleep, 0)
```

File: switchio/loops/proactor.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                    cbs = callbacks.get(evname, ())
+                    self.log.debug(
+                        "consumer '{}' has callback {} registered for ev {}"
+                        .format(cid, cbs, evname)
+                    )
+                    # look up the client's callback chain and run
+                    # e -> handler -> cb1, cb2, ... cbN
+                    # XXX assign ret on each interation in an attempt to avoid
+                    # python's dynamic scope lookup
+                    for cb, ret in zip(cbs, itertools.repeat(ret)):
```
Comment (vodik) [inline switchio/loops/proactor.py]: Might want to look at using `itertools.izip` for python 2, depending how much you care about that platform at this point.

File: switchio/loops/proactor.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+                self.log.warning("Caught ESL error for event '{}':\n{}"
+                                 .format(evname, traceback.format_exc()))
+            except Exception:
+                self.log.exception(
+                    "Failed to process event {} with uid {}"
+                    .format(evname, uid)
+                )
+            return consumed
+        else:
+            self.log.error("Unknown event '{}'".format(evname))
```
Comment (vodik) [inline switchio/loops/proactor.py]: You should turn on flake8's cyclomatic complexity measurements 😉

File: switchio/loops/proactor.py (hunk 3 above)
Comment (tgoodlet) [reply switchio/loops/proactor.py]: Yeah i want to make it obvious that this is a ridiculous thing to expect users of `asyncio` to do.
I'd like to possibly talk to the cpython guys and get a proper future api for this.

File: switchio/loops/proactor.py (hunk 5 above)
Comment (tgoodlet) [reply switchio/loops/proactor.py]: Ok i'll make an issue to investigate.

**Pull request merged**